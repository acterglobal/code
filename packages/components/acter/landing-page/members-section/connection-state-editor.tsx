import React, { FC } from 'react'

import {
  Box,
  IconButton,
  FormControl,
  MenuItem,
  Select,
  createStyles,
  withStyles,
} from '@material-ui/core'
import { Save, Cancel } from '@material-ui/icons'

import { Formik, Form } from 'formik'

import { LoadingSpinner } from '@acter/components/util/loading-spinner'
import { useUpdateActerConnection } from '@acter/lib/acter/use-update-connection'
import { ActerConnection, ActerConnectionRole } from '@acter/schema'

export interface ConnectionStateEditorProps {
  connection: ActerConnection
  onCancel: () => void
}

export const ConnectionStateEditor: FC<ConnectionStateEditorProps> = ({
  connection,
  onCancel,
}) => {
  const [
    updateActerConnection,
    { loading: updatingConnection },
  ] = useUpdateActerConnection()

  const { Follower, role } = connection
  const initialValues = {
    role,
  }
  const handleSubmit = ({ role }) => updateActerConnection(connection, role)

  if (updatingConnection) return <LoadingSpinner />

  return (
    <Formik initialValues={initialValues} onSubmit={handleSubmit}>
      {({ dirty, values, handleChange, handleReset, submitForm }) => {
        const handleCancel = () => {
          handleReset()
          onCancel()
        }
        return (
          <Form>
            <FormControl>
              <Container>
                <Select name="role" value={values.role} onChange={handleChange}>
                  {Object.keys(ActerConnectionRole).map((key) => (
                    <MenuItem
                      key={`acter-${Follower.id}-connection-picker-${key}`}
                      value={key}
                    >
                      {ActerConnectionRole[key]}
                    </MenuItem>
                  ))}
                </Select>
                <IconButton disabled={!dirty} onClick={submitForm}>
                  <Save />
                </IconButton>
                <IconButton onClick={handleCancel}>
                  <Cancel />
                </IconButton>
              </Container>
            </FormControl>
          </Form>
        )
      }}
    </Formik>
  )
}

const Container = withStyles(() =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
    },
  })
)(Box)
