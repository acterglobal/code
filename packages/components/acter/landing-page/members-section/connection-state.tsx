import React, { FC, useState, useEffect } from 'react'
import { Box, IconButton, createStyles, withStyles } from '@material-ui/core'
import { Create as EditIcon } from '@material-ui/icons'
import {
  ConnectionStateEditor,
  ConnectionStateEditorProps,
} from '@acter/components/acter/landing-page/members-section/connection-state-editor'
import { ActerConnection } from '@acter/schema'

export interface ConnectionStateProps
  extends Omit<ConnectionStateEditorProps, 'onCancel'> {
  connection: ActerConnection
  canEdit: boolean
}

export const ConnectionState: FC<ConnectionStateProps> = ({
  connection,
  canEdit,
  onSubmit,
}) => {
  const [isEditing, setIsEditing] = useState(false)

  useEffect(() => {
    setIsEditing(false)
  }, [connection])

  if (canEdit && isEditing) {
    return (
      <ConnectionStateEditor
        connection={connection}
        onCancel={() => setIsEditing(false)}
        onSubmit={onSubmit}
      />
    )
  }
  return (
    <Container>
      {connection.role}
      {canEdit && (
        <IconButton onClick={() => setIsEditing(true)}>
          <EditIcon />
        </IconButton>
      )}
    </Container>
  )
}

const Container = withStyles(() =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
    },
  })
)(Box)
