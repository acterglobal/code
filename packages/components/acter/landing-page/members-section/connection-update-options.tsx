import React, { FC } from 'react'

import {
  IconButton,
  createStyles,
  makeStyles,
  Theme,
  MenuItem,
} from '@material-ui/core'
import { MoreHoriz as DotsIcon } from '@material-ui/icons'

import { DropdownMenu } from '@acter/components/util/dropdown-menu'
import { useUpdateActerConnection } from '@acter/lib/acter/use-update-connection'
import { ActerConnection, ActerConnectionRole } from '@acter/schema'

const { ADMIN, MEMBER, REJECTED } = ActerConnectionRole
export interface ConnectionStateProps {
  connection: ActerConnection
  canEdit: boolean
}

export const ConnectionUpdateOptions: FC<ConnectionStateProps> = ({
  connection,
  canEdit,
}) => {
  const classes = useStyles()

  if (!canEdit) return null

  return (
    <DropdownMenu
      anchorNode={
        <IconButton disableRipple className={classes.iconButton}>
          <DotsIcon fontSize="small" />
        </IconButton>
      }
    >
      {connection.role === ADMIN && (
        <>
          <ActionItem
            label="Remove as Admin"
            connection={connection}
            updateRole={MEMBER}
          />
          <ActionItem
            label="Remove as Member"
            connection={connection}
            updateRole={REJECTED}
          />
        </>
      )}
      {connection.role === MEMBER && (
        <>
          <ActionItem
            label="Make as Admin"
            connection={connection}
            updateRole={ADMIN}
          />
          <ActionItem
            label="Remove as Member"
            connection={connection}
            updateRole={REJECTED}
          />
        </>
      )}
    </DropdownMenu>
  )
}

type ActionItemProps = {
  label: string
  connection: ActerConnection
  updateRole: ActerConnectionRole
}
const ActionItem: FC<ActionItemProps> = ({ label, connection, updateRole }) => {
  const classes = useStyles()
  const [_, updateConnection] = useUpdateActerConnection()

  const handleClick = () => updateConnection(connection, updateRole)

  return (
    <MenuItem className={classes.menuItem} onClick={handleClick}>
      {label}
    </MenuItem>
  )
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    iconButton: {
      width: theme.spacing(6),
      backgroundColor: theme.colors.grey.extraLight,
      borderRadius: 5,
      height: theme.spacing(3.5),
    },
    menuItem: {
      margin: theme.spacing(1),
      fontSize: '0.8rem',
      fontWeight: theme.typography.fontWeightMedium,
      color: theme.palette.secondary.main,
      '&:hover': {
        backgroundColor: theme.colors.white,
      },
    },
  })
)
