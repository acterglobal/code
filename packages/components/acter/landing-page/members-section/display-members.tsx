import React, { FC } from 'react'
import pluralize from 'pluralize'
import {
  Box,
  Divider,
  Typography,
  List,
  ListItem,
  ListItemAvatar,
  ListItemText,
  ListItemSecondaryAction,
} from '@material-ui/core'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { userHasRoleOnActer } from '@acter/lib/user/user-has-role-on-acter'
import { ActerAvatar } from '@acter/components/acter/avatar'
import {
  ConnectionState,
  ConnectionStateProps,
} from '@acter/components/acter/landing-page/members-section/connection-state'
import {
  Acter,
  ActerConnection,
  ActerConnectionRole,
  User,
} from '@acter/schema/types'

import { MemberType } from '@acter/lib/constants'
export interface DisplayMembersProps {
  /**
   * The acter on which we are viewing members
   */
  acter: Acter
  /**
   * The list of acters we are displaying
   */
  followers: ActerConnection[]
  /**
   * The currently logged in user
   */
  user: User
  /**
   * The type of connection
   */
  type: MemberType
  /**
   * Action when Member state changes
   */
  onConnectionStateChange: ConnectionStateProps['onSubmit']
}

export const DisplayMembers: FC<DisplayMembersProps> = ({
  acter,
  user,
  followers = [],
  type,
  onConnectionStateChange,
}) => {
  const classes = useStyles()

  const showJoinState = userHasRoleOnActer(
    user,
    ActerConnectionRole.MEMBER,
    acter
  )
  const canEdit = userHasRoleOnActer(user, ActerConnectionRole.ADMIN, acter)

  return (
    <Box className={classes.container}>
      <Typography className={classes.heading} variant="h5">
        {followers.length} {pluralize(type)}
      </Typography>
      <Divider />

      <List className={classes.members}>
        {followers.map((connection) => {
          const { Follower } = connection
          return (
            <>
              <ListItem>
                <ListItemAvatar>
                  <ActerAvatar acter={Follower} />
                </ListItemAvatar>
                <ListItemText
                  classes={{
                    primary: classes.name,
                    secondary: classes.acterType,
                  }}
                  className={classes.memberInfo}
                  primary={Follower.name}
                  secondary={Follower.ActerType.name}
                />
                {showJoinState && (
                  <ListItemSecondaryAction>
                    <ConnectionState
                      acter={acter}
                      connection={connection}
                      canEdit={canEdit && Follower.id !== user.Acter.id}
                      onSubmit={onConnectionStateChange}
                    />
                  </ListItemSecondaryAction>
                )}
              </ListItem>
              <Divider
                classes={{ root: classes.divider }}
                variant="inset"
                component="li"
              />
            </>
          )
        })}
      </List>
    </Box>
  )
}

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    height: '100%',
    padding: theme.spacing(3),
    backgroundColor: theme.palette.background.paper,
    borderRadius: theme.spacing(1),
    borderColor: theme.palette.divider,
    borderWidth: 'thin',
    borderStyle: 'solid',
    overflow: 'hidden',
  },
  heading: {
    fontSize: '0.9rem',
    fontWeight: theme.typography.fontWeightBold,
    color: theme.colors.grey.dark,
    marginBottom: 20,
  },
  members: {
    height: '100%',
    overflowY: 'scroll',
    width: '100%',
  },
  memberInfo: {
    marginLeft: 30,
    fontWeight: theme.typography.fontWeightBold,
    textTransform: 'capitalize',
  },
  name: {
    fontWeight: theme.typography.fontWeightBold,
    color: theme.colors.grey.dark,
  },
  acterType: {
    color: theme.colors.grey.dark,
  },
  divider: {
    marginLeft: 100,
  },
}))
