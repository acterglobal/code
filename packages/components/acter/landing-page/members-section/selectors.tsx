import React, { FC } from 'react'
import { Box, Button } from '@material-ui/core'
import { green } from '@material-ui/core/colors'
import { makeStyles, Theme } from '@material-ui/core/styles'
import clsx from 'clsx'
import { justCapitalize } from '@acter/components/util/just-capitalize'

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    display: 'flex',
    justifyContent: 'space-between',
    marginBottom: theme.spacing(1),
    paddingLeft: theme.spacing(1),
  },
  button: {
    height: 25,
    minWidth: 120,
    borderRadius: theme.spacing(3),
    marginRight: theme.spacing(1),
    color: theme.palette.secondary.main,
    backgroundColor: 'white',
    fontSize: '0.7rem',
    '&:hover': {
      backgroundColor: 'white',
    },
  },
  active: {
    backgroundColor: green[400],
    color: 'white',
    '&:hover': {
      backgroundColor: green[600],
    },
  },
}))

export interface SelectorProps {
  selectors: string[]
  activeSelector: string
  onChange: (selector: string) => void
}

export const Selectors: FC<SelectorProps> = ({
  selectors,
  activeSelector,
  onChange,
}) => {
  const classes = useStyles()

  return (
    <Box className={classes.container}>
      <Box role="tablist">
        {selectors.map((selector) => (
          <Button
            key={`members-selector-${selector}`}
            role="tab"
            className={clsx(
              classes.button,
              activeSelector === selector && classes.active
            )}
            variant="contained"
            onClick={() => onChange(selector)}
          >
            {selector && justCapitalize(selector)}
          </Button>
        ))}
      </Box>
    </Box>
  )
}
