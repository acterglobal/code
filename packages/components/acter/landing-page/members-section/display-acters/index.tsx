import React, { FC } from 'react'

import { useRouter } from 'next/router'

import { List, Theme, ListItem, ListItemSecondaryAction, Button } from '@mui/material';

import createStyles from '@mui/styles/createStyles';
import makeStyles from '@mui/styles/makeStyles';

import { MemberDetails } from '@acter/components/acter/landing-page/members-section/display-members/member-details'
import { acterAsUrl } from '@acter/lib/acter/acter-as-url'
import { ActerConnection } from '@acter/schema'

interface DisplayActersProps {
  followers: ActerConnection[]
}
export const DisplayActers: FC<DisplayActersProps> = ({ followers = [] }) => {
  const classes = useStyles()
  const router = useRouter()

  return (
    <List className={classes.list}>
      {followers.map(({ Follower }) => (
        <ListItem>
          <MemberDetails follower={Follower} />
          <ListItemSecondaryAction>
            <Button
              className={classes.button}
              onClick={() => router.push(acterAsUrl({ acter: Follower }))}
            >
              View
            </Button>
          </ListItemSecondaryAction>
        </ListItem>
      ))}
    </List>
  )
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    list: {
      paddingLeft: theme.spacing(6.5),
      paddingRight: theme.spacing(6.5),
    },
    button: {
      backgroundColor: theme.colors.grey.extraLight,
      textTransform: 'capitalize',
      color: theme.palette.secondary.main,
      fontSize: 12,
      width: theme.spacing(8.5),
    },
  })
)
