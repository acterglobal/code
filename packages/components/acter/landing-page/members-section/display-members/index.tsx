import React, { FC } from 'react'
import { di } from 'react-magnetic-di'

import { Box, Divider, Typography, List } from '@material-ui/core'
import { makeStyles, Theme } from '@material-ui/core/styles'

import pluralize from 'pluralize'

import { DisplayMemberItem } from '@acter/components/acter/landing-page/members-section/display-members/display-member-item'
import { Link } from '@acter/components/util/anchor-link'
import { LoadingSpinner } from '@acter/components/util/loading-spinner'
import { acterAsUrl } from '@acter/lib/acter/acter-as-url'
import { useActer } from '@acter/lib/acter/use-acter'
import { MemberType } from '@acter/lib/constants'
import { useUser } from '@acter/lib/user/use-user'
import { userHasRoleOnActer } from '@acter/lib/user/user-has-role-on-acter'
import { ActerConnection, ActerConnectionRole } from '@acter/schema'

export interface DisplayMembersProps {
  /**
   * The list of acters we are displaying
   */
  followers: ActerConnection[]
  /**
   * The type of connection
   */
  type: MemberType
  /**
   * Boolean to indicate organisation member/follower type
   */
  isOrganisation?: boolean
}

export const DisplayMembers: FC<DisplayMembersProps> = ({
  followers = [],
  type,
  isOrganisation,
}) => {
  di(useActer, useUser)
  const classes = useStyles()
  const { acter, loading: acterLoading } = useActer()
  const { user, loading: userLoading } = useUser()

  if (acterLoading || userLoading) return <LoadingSpinner />
  if (!acter) return null

  const showJoinState = userHasRoleOnActer(
    user,
    ActerConnectionRole.MEMBER,
    acter
  )
  const canEdit = userHasRoleOnActer(user, ActerConnectionRole.ADMIN, acter)

  return (
    <Box className={classes.container}>
      <Typography className={classes.heading} variant="h5">
        {followers.length} {pluralize(type)}
      </Typography>
      <Divider />

      <List className={classes.members}>
        {followers.map((connection) => {
          const { Follower } = connection

          if (isOrganisation) {
            return (
              <Link
                href={`${acterAsUrl({ acter: Follower })}`}
                key={`follower-${Follower.id}`}
              >
                <DisplayMemberItem
                  user={user}
                  Follower={Follower}
                  connection={connection}
                  showJoinState={showJoinState}
                  canEdit={canEdit}
                />
              </Link>
            )
          }

          return (
            <DisplayMemberItem
              key={`follower-${Follower.id}`}
              user={user}
              Follower={Follower}
              connection={connection}
              showJoinState={showJoinState}
              canEdit={canEdit}
            />
          )
        })}
      </List>
    </Box>
  )
}

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    height: '100%',
    padding: theme.spacing(3),
    backgroundColor: theme.palette.background.paper,
    borderRadius: theme.spacing(1),
    borderColor: theme.palette.divider,
    borderWidth: 'thin',
    borderStyle: 'solid',
    overflow: 'hidden',
  },
  heading: {
    fontSize: '0.9rem',
    fontWeight: theme.typography.fontWeightBold,
    color: theme.colors.grey.dark,
    marginBottom: 20,
  },
  members: {
    height: '100%',
    overflowY: 'scroll',
    width: '100%',
  },
}))
