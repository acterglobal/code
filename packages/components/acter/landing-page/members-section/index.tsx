import React, { FC, useState } from 'react'
import { di } from 'react-magnetic-di/macro'

import { Box } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'

import { DisplayMembers } from '@acter/components/acter/landing-page/members-section/display-members'
import { Selectors } from '@acter/components/acter/landing-page/members-section/selectors'
import { LoadingSpinner } from '@acter/components/util/loading-spinner'
import { mapFollowersByType } from '@acter/lib/acter/map-followers-by-type'
import { useActer } from '@acter/lib/acter/use-acter'
import { MemberType } from '@acter/lib/constants'

const { ORGANISATIONS, PEOPLE } = MemberType

const useStyles = makeStyles({
  container: {
    minWidth: 400,
  },
})

export const MembersSection: FC = () => {
  di(useActer)
  const classes = useStyles()
  const [activeSelector, setActiveSelector] = useState<MemberType>(PEOPLE)

  const { acter, loading: acterLoading } = useActer()

  if (acterLoading) return <LoadingSpinner />
  if (!acter) return null

  const followers = mapFollowersByType(acter)

  const handleSelectorChange = (selector) => {
    setActiveSelector(selector)
  }

  return (
    <Box className={classes.container}>
      <Selectors
        selectors={[PEOPLE, ORGANISATIONS]}
        activeSelector={activeSelector}
        onChange={handleSelectorChange}
      />
      <DisplayMembers
        followers={
          activeSelector === PEOPLE ? followers.user : followers.organisation
        }
        type={activeSelector}
        isOrganisation={activeSelector === ORGANISATIONS ? true : false}
      />
    </Box>
  )
}
