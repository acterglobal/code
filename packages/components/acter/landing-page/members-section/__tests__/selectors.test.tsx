/**
 * @jest-environment jsdom
 */
import React from 'react'

import { Selectors } from '@acter/components/acter/landing-page/members-section/selectors'
import { useTranslation } from '@acter/lib/i18n/use-translation'
import { render, screen } from '@acter/lib/test-utils'

jest.mock('@acter/lib/i18n/use-translation')

const FOO = 'Foo'
const BAR = 'Bar'
const selectors = [FOO, BAR]
const onChange = () => null

describe('<Selectors>', () => {
  const mockUseTranslation = useTranslation as jest.Mock

  beforeEach(() => {
    mockUseTranslation.mockReturnValue({ t: (key) => key })
  })

  it('should display a button for each selector', async () => {
    render(
      <Selectors
        selectors={[FOO, BAR]}
        activeSelector={FOO}
        totalResults={0}
        onChange={onChange}
      />
    )
    expect(screen.getByRole('tablist')).toBeTruthy()
    const tabs = await screen.findAllByRole('tab')
    expect(tabs).toHaveLength(2)
    expect(tabs[0].textContent).toBe(FOO)
    expect(tabs[1].textContent).toBe(BAR)
  })

  it('should display number of results', async () => {
    render(
      <Selectors
        selectors={[FOO, BAR]}
        activeSelector={FOO}
        totalResults={10}
        onChange={onChange}
      />
    )

    expect(screen.getByRole('heading', { name: '10 Foos' })).toBeTruthy()
  })

  it('should highlight the selected tab', () => {
    const { asFragment } = render(
      <Selectors
        selectors={selectors}
        activeSelector={FOO}
        totalResults={10}
        onChange={onChange}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
