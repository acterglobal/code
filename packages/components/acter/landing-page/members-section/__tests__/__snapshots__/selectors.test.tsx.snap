// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selectors> should highlight the selected tab 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-container-9 MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
      role="tablist"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root makeStyles-button-10 makeStyles-active-11 css-1qk2aq7-MuiButtonBase-root-MuiButton-root"
        role="tab"
        tabindex="0"
        type="button"
      >
        Foo
      </button>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root makeStyles-button-10 css-1qk2aq7-MuiButtonBase-root-MuiButton-root"
        role="tab"
        tabindex="0"
        type="button"
      >
        Bar
      </button>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 makeStyles-results-12 css-1hivc1e-MuiTypography-root"
        role="heading"
      >
        10 Foos
      </p>
    </div>
  </div>
</DocumentFragment>
`;
